internal/plugin/server.go

```diff
@@ -69,41 +66,25 @@ type NvidiaDevicePlugin struct {
 	mpsHostRoot mps.Root
 }
 
-// NewNvidiaDevicePlugin returns an initialized NvidiaDevicePlugin
-func NewNvidiaDevicePlugin(config *spec.Config, kubeletSocket string, resourceManager rm.ResourceManager, cdiHandler cdi.Interface) (*NvidiaDevicePlugin, error) {
-	_, name := resourceManager.Resource().Split()
-
-	deviceListStrategies, _ := spec.NewDeviceListStrategies(*config.Flags.Plugin.DeviceListStrategy)
-
-	pluginName := "nvidia-" + name
-	pluginPath := filepath.Join(pluginapi.DevicePluginPath, pluginName)
-
-	var mpsDaemon *mps.Daemon
-	var mpsHostRoot mps.Root
-	if config.Sharing.SharingStrategy() == spec.SharingStrategyMPS {
-		// TODO: It might make sense to pull this logic into a resource manager.
-		for _, device := range resourceManager.Devices() {
-			if device.IsMigDevice() {
-				return nil, errors.New("sharing using MPS is not supported for MIG devices")
-			}
-		}
-		mpsDaemon = mps.NewDaemon(resourceManager, mps.ContainerRoot)
-		mpsHostRoot = mps.Root(*config.Flags.CommandLineFlags.MpsRoot)
+// devicePluginForResource creates a device plugin for the specified resource.
+func (o *options) devicePluginForResource(resourceManager rm.ResourceManager) (Interface, error) {
+	mpsDaemon, mpsHostRoot, err := o.getMPSDaemon(resourceManager)
+	if err != nil {
+		return nil, err
 	}
```

Comment:
Can we move the if check for `o.config.Sharing.SharingStrategy() != spec.SharingStrategyMPS ` out here to make it clear that we only actually set these variables to anything if MPS is configured. Alternatively, rename the function something like `getMPSDaemonIfConfigured()`? As it reads now it seems as if we are *always* getting an MPS server setup for the plugin.

Reply:
I'm more partial to the latter. I've reworked it a bit by introducing an `mpsOptions` type and moved the MPS-specifics to a separate file.

We now set `enabled` on this type specifically instead of having to query the sharing strategy all the time.